test/hotspot/jtreg/runtime/cds/appcds/aotClassLinking/WeakReferenceTest.java

```diff
@@ -0,0 +1,309 @@
+
+        if (isProduction && Inner.WeakReferenceTestApp_clinit_executed) {
+            throw new RuntimeException("WeakReferenceTestApp should have been aot-inited");
+        }
+
+        if (isProduction) {
+            // A GC should have happened before the heap objects are written into
+            // the AOT cache. So any unreachable referents should have been collected.
+        } else {
+            // We are in the training run. Simulate the GC mentioned in the above comment,
+            // so the test cases should observe the same states as in the production run.
+            System.gc();
+        }
+
+        testCollectedInAssembly(isProduction);
+        testWeakReferenceCollection(isProduction);
+    }
+
+    //----------------------------------------------------------------------
+    // Set up for testCollectedInAssembly()
+    static WeakReference refToCollectedObj;
+
+    static void aot_init_for_testCollectedInAssembly() {
+        // The referent will be GC-ed in the assembly run when the JVM forces a full GC.
+        refToCollectedObj = new WeakReference(new String("collected in assembly"));
+    }
+
+    // [TEST CASE] Test the storage of a WeakReference whose referent has been collected during the assembly phase.
+    static void testCollectedInAssembly(boolean isProduction) {
+        System.out.println("refToCollectedObj.get() = " + refToCollectedObj.get());
+
+        if (refToCollectedObj.get() != null) {
+            throw new RuntimeException("refToCollectedObj.get() should have been GC'ed");
+        }
+    }
+
+    //----------------------------------------------------------------------
+    // Set up for testWeakReferenceCollection()
+    static Object root;
+    static WeakReference ref;
+
+    static void aot_init_for_testWeakReferenceCollection() {
+        root = new String("to be collected in production");
+        ref = makeRef();
+    }
+
+    static WeakReference makeRef() {
+        System.out.println("WeakReferenceTestApp::makeRef() is executed");
+        WeakReference r = new WeakReference(root);
+        System.out.println("r.get() = " + r.get());
+
+        ShouldNotBeAOTInited.doit();
+        return r;
+    }
+
+    static WeakReference makeRef2() {
+        return new WeakReference(new String("to be collected in production"));
+    }
+
+
+    // [TEST CASE] A WeakReference allocated in assembly phase should be collectable in the production run
+    static void testWeakReferenceCollection(boolean isProduction) {
+        WeakReference ref2 = makeRef2();
+        System.out.println("ref.get() = " + ref.get());   // created during assembly phase
+        System.out.println("ref2.get() = " + ref2.get()); // created during production run
+
+        if (ref.get() == null) {
+            throw new RuntimeException("ref.get() should not be null");
+        }
+        if (ref2.get() == null) {
+            throw new RuntimeException("ref2.get() should not be null");
+        }
```

Comment:
The referent of ref2 is never strong. If a GC occurs between makeRef2() and this check, we could hit this Exception. The test should either keep it strongly reachable until after this check or just not test wether it is reachable before the call to System.gc().

Reply:
I removed the check that was before the GC.